gitindex/index.go

```diff
@@ -532,6 +539,229 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
 	return ignore.ParseIgnoreFile(strings.NewReader(content))
 }
 
+type gitIndexConfig struct {
+	// prepareDeltaBuildMetadata, if not nil, is the function is used to calculate the metadata that will be used to
+	// prepare the build.Builder instance for generating a delta build.
+	//
+	// If prepareDeltaBuildMetadata is nil, gitindex.prepareDeltaBuild will be used instead.
+	prepareDeltaBuildMetadata func(options Options, repository *git.Repository) (repos map[fileKey]BlobLocation, branchMap map[fileKey][]string, branchVersions map[string]map[string]plumbing.Hash, changedOrDeletedPaths []string, err error)
+
+	// prepareNormalBuildMetadata, if not nil, is the function is used to calculate the metadata that will be used to
+	// prepare the build.Builder instance for generating a normal build.
+	//
+	// If prepareNormalBuildMetadata is nil, gitindex.prepareNormalBuild will be used instead.
+	prepareNormalBuildMetadata func(options Options, repository *git.Repository) (repos map[fileKey]BlobLocation, branchMap map[fileKey][]string, branchVersions map[string]map[string]plumbing.Hash, err error)
+}
+
+func prepareDeltaBuild(options Options, repository *git.Repository) (repos map[fileKey]BlobLocation, branchMap map[fileKey][]string, branchVersions map[string]map[string]plumbing.Hash, changedOrDeletedPaths []string, err error) {
```

Comment:
The lack of branchVersions support is just a thing that we chose to de-prioritize for the prototype. I think it's fine to keep it in the signature, since we definitely want to support it in the full implementation?